server/src/main/java/org/elasticsearch/gateway/PersistedClusterStateService.java

```diff
@@ -716,28 +706,39 @@ private WriterStats overwriteMetadata(Metadata metadata) throws IOException {
+            try (DocumentBuffer documentBuffer = allocateBuffer()) {
+
+                final Document globalMetadataDocument = makeGlobalMetadataDocument(metadata, documentBuffer);
                 for (MetadataIndexWriter metadataIndexWriter : metadataIndexWriters) {
-                    metadataIndexWriter.updateGlobalMetadata(globalMetadataDocument.getDocument());
+                    metadataIndexWriter.updateGlobalMetadata(globalMetadataDocument);
                 }
-            }
 
-            for (ObjectCursor<IndexMetadata> cursor : metadata.indices().values()) {
-                final IndexMetadata indexMetadata = cursor.value;
-                try (ReleasableDocument indexMetadataDocument = makeIndexMetadataDocument(indexMetadata)) {
+                for (ObjectCursor<IndexMetadata> cursor : metadata.indices().values()) {
+                    final IndexMetadata indexMetadata = cursor.value;
+                    final Document indexMetadataDocument = makeIndexMetadataDocument(indexMetadata, documentBuffer);
                     for (MetadataIndexWriter metadataIndexWriter : metadataIndexWriters) {
-                        metadataIndexWriter.updateIndexMetadataDocument(indexMetadataDocument.getDocument(), indexMetadata.getIndex());
+                        metadataIndexWriter.updateIndexMetadataDocument(indexMetadataDocument, indexMetadata.getIndex());
                     }
                 }
-            }
 
-            // Flush, to try and expose a failure (e.g. out of disk space) before committing, because we can handle a failure here more
-            // gracefully than one that occurs during the commit process.
-            for (MetadataIndexWriter metadataIndexWriter : metadataIndexWriters) {
-                metadataIndexWriter.flush();
+                documentBufferUsed = documentBuffer.getMaxUsed();
+
+                // Flush, to try and expose a failure (e.g. out of disk space) before committing, because we can handle a failure here more
+                // gracefully than one that occurs during the commit process.
+                for (MetadataIndexWriter metadataIndexWriter : metadataIndexWriters) {
+                    metadataIndexWriter.flush();
+                }
+
+                return new WriterStats(true, metadata.indices().size(), 0);
             }
+        }
 
-            return new WriterStats(true, metadata.indices().size(), 0);
+        private DocumentBuffer allocateBuffer() {
+            // heuristics for picking the initial buffer size based on the buffer we needed last time: try and fit within a single page,
+            // but if we needed more than a single page last time then allow a bit more space to try and avoid needing to grow the buffer
+            // later on.
+            final int extraSpace = documentBufferUsed <= PageCacheRecycler.PAGE_SIZE_IN_BYTES ? 0 : PageCacheRecycler.PAGE_SIZE_IN_BYTES;
```

Comment:
This is something of an educated guess TBH. An extra 16kB on top of whatever we needed last time doesn't seem so bad, we could also use something like `min(16kB, documentBufferUsed/8)` to make it a bit more proportional without blowing up on enormous states but whatever we do here is almost certainly good enough.